### DISCUSSION TITLE
Respect HTTP Accept header

### DISCUSSION THREAD

**Discussion by sergiodxa** [remix-run/remix#5024] [Proposals]
# Proposal

Make Remix respect the HTTP Accept header and if it's not text/html send the loader data instead, even if the route is a UI route.

## Background

The Accept header let clients (like browsers or another app) let the server know what type of responses it can handle. Then the server can send a different response type based on that.

Right now, you can use it in a resource route quite simple:

```ts
export async function loader({ request }: LoaderArgs) {
  let accept = request.headers.get("Accept")
  let data = await getDataSomehow()
  if (accept.includes("application/json")) return json(data);
  if (accept.includes("application/pdf")) return pdf(renderToPDF(data));
  throw badRequest("Invalid accept type") // or fallback to json
}
```

If a UI route (a route with a `export default`) wants to support this it's not possible, Remix will always attempt to render the app, the only way to skip it is with the special `_data` search param.

## Use Case

- Let developer use the same route loader without creating a separate resource route
- Improve web standards support
- Allow the same URL to send different representations of the same data (e.g. HTML, JSON, XML or PDF)

## API/Examples

The idea would be for Remix to detect the header and if the Accept is not `text/html` treat the request as a resource route request.

This would mean that Remix will call any leaf route matching the request and it will send the loader response, if the developer wants to support more than JSON it should manually check the Accept header to return something different based on that.

If the request comes with the `text/html` inside Accept Remix should treat the request a document one and call all matching routes.

Remix must still use the `_data` search param to ensure the URL for JSON and the URL for document made by Remix itself are different and avoid any issue with CDNs not respecting Accept header.

For requests with the Accept not being `text/html` it should be documented somewhere that CDNs may cache the responses so developers can take care of that and avoid setting Cache-Control or use a CDN which correctly handle Accept and Vary.

## Prior Art

Rails works this way, you can use a single controller to send different response types like HTML or JSON, they even give you a function to do the conditions easier

```rb
respond_to do |format|
  # Rails take care of using the correct template, Remix this will happen outside the loader
  format.html
  # You could send an XML response
  format.xml  { render xml: @users }
  # Our you could send a JSON response
  format.json { render json: @users }
end
```

This could either be implemented by Remix or left to devs to do it, maybe Remix Utils could provide a helper

```ts
let data = await getDataSomehow();
return respondTo(request.headers, {
  json() {
    return json(data);
  },
  xml() {
    let string = createXMLFromJSON(data);
    return xml(string);
  },
  pdf() {
    let pdfFile = renderPDF(data);
    return pdf(pdfFile);
  },
  "application/custom-type": () => {
    return new Response(parseDataSomehow(data), {
      headers: { "Content-Type": "application/custom-type" }
    });
  },
  default() {
    // set a default for unknown Accept values, you could throw here or return another one
    return json(data);
  }
});
```

**Comment by machour** [comment]
This was discussed in [Roadmap #5, around 20'55](https://www.youtube.com/live/tdVFZidXGZo?feature=share&t=1255)
Ryan and Michael are incline to *not* doing that, because of CDN caching, and avoidable "complexity".

The workaround is to implement a `myroute[.ext].ts` routes, and serve that different version of content from there, or continue to use the _data param.

**Reply by sergiodxa** [reply]
That workaround doesn't work if you want to migrate a current app to Remix without breaking URLs, also

> Remix must still use the _data search param to ensure the URL for JSON and the URL for document made by Remix itself are different and avoid any issue with CDNs not respecting Accept header.

I mentioned that Remix should still use `_data` search param to avoid this CDN caching issue, but supporting this can still be useful for app developers who know their CDN either supports Accept header or they're not caching at the CDN so this is not an issue for them.

**Reply by jamesarosen** [reply]
> I mentioned that Remix should still use _data search param to avoid this CDN caching issue

Some CDNs (notably Amazon CloudFront) do not vary on query-params by default and others will refuse to cache anything that has query-params. Rails calls this out in their [static assets guide](https://guides.rubyonrails.org/asset_pipeline.html#what-is-fingerprinting-and-why-should-i-care-questionmark):

> The query string strategy has several disadvantages:
> Not all caches will reliably cache content where the filename only differs by query parameters. [Steve Souders recommends](https://www.stevesouders.com/blog/2008/08/23/revving-filenames-dont-use-querystring/), "...avoiding a querystring for cacheable resources". He found that in this case 5-20% of requests will not be cached. Query strings in particular do not work at all with some CDNs for cache invalidation.

**Reply by sergiodxa** [reply]
Using that search param is how it works right now

**Comment by jamesarosen** [comment]
Building on this topic, the [Form Validation docs](https://remix.run/docs/en/v1/guides/data-writes#form-validation) show this:

```js
export const action = async ({ request }: ActionArgs) => {
  const formData = await request.formData();
  const [errors, project] = await createProject(formData);

  if (errors) {
    const values = Object.fromEntries(formData);
    return json({ errors, values });
  }

  return redirect(`/projects/${project.id}`);
};
```

That absolutely works, but it delivers a pretty lousy experience for non-JS browsers. Who wants to see a page with nothing but a JSON-encoded error message?

Adapting the Rails pattern mentioned above, this might become
```js
return respondTo(request.headers, {
  json: () => json({ errors, values }),
  html: () => {
    return redirectBackOrDefault('/', {
      headers: { 'set-cookie': flash(errors) }
    }
  },
})
```

**Reply by machour** [reply]
@jamesarosen if a user don't have JS enabled, the form will trigger a full page reload and the json() data will correctly populate the page.

**Reply by jamesarosen** [reply]
> the form will trigger a full page reload

This form

```html
<form method="POST" action="/projects/123">
  <button type="submit">Save</button>
</form>
```

will certainly trigger a full page load, but the page it loads will be the response to `POST /projects/123`, which will be JSON from

```ts
return json({ errors, values });
```

In order to re-render the form with an error message, we would need to change the action handler to save the errors in session flash and redirect:

```ts
if (errors) {
  session.flash("formErrors", JSON.stringify(errors));
}

redirect(`/projects/${project.id}`);
```

**Reply by sergiodxa** [reply]
If `/projects/123` is a resource route then yes, you will need a redirect, otherwise Remix should render the component of `/projects/123` when using a `<form>` instead of `<Form>`.

It's also possible to know if a fetcher.Form was submitted with or without JS and change the response from json to a redirect, check https://sergiodxa.com/articles/progressively-enhance-the-usefetcher-hook-in-remix.

If this is not a case, try to create a reproduction repo because this may be a bug.

**Reply by jamesarosen** [reply]
> Remix should render the component of `/projects/123` when using a `<form>` instead of `<Form>`.

I think this is the key disconnect. I'm glad you put your finger on it. Yes, Remix's front-end components enrich this behavior. I was thinking of the HTTP APIs in isolation, not necessarily consumed by the Remix UI.

**Comment by kael** [comment]
FYI, [SvelteKit supports content-negotiation](https://kit.svelte.dev/docs/routing#server-content-negotiation) by using two different files formats under the same route :

> To determine which, SvelteKit applies the following rules:
> 
>     * PUT/PATCH/DELETE/OPTIONS requests are always handled by +server.js since they do not apply to pages
>
>     * GET/POST/HEAD requests are treated as page requests if the accept header prioritises text/html 
>       (in other words, it's a browser page request), else they are handled by +server.js.
>
>     * Responses to GET requests will include a Vary: Accept header, so that proxies and browsers cache HTML and  
>       JSON responses separately.

**Comment by tmcw** [comment]
👋 Another vote in support of this. My usecase is pretty similar to how deno.land supports a page like this: https://deno.land/std@0.199.0/bytes/bytes_list.ts - which renders HTML for browsers but returns TypeScript for Deno and other clients.

The options for this kind of problem - content negotiation - in Remix are lackluster. I can't implement a system where you add `.ts` to the end of URLs to get the TypeScript version because the router doesn't support regex routes and so can't differentiate between `/route/$param.ts` and `/route/$param[.ts].ts`. And I can't support a way to redirect away from `/route/$param.ts` _to_ a specific TypeScript-serving resource route, because as this discussion mentions, it's impossible to have a UI route that does content negotiation.

**Reply by kael** [reply]
> it's impossible to have a UI route that does content negotiation.

I've used a routes `redirect` from `/user/$id.js` to `/user/$id[.json].js` [for that case](https://github.com/remix-run/remix/discussions/5464#discussioncomment-4996084).

**Comment by nickretallack** [comment]
TIL that the behavior of an `action` changes based on whether on not there is a `default` export in the same route file.  This seems like an unfortunate counter-intuitive gotcha.  Please consider changing this, or you'll have more people tripping on this problem and getting upset with the framework.

The `_data` parameter work-around seems very bad.  The information you pass to the `_data` parameter is redundant with the URL that's already in the request.  What happens if those two don't match?  Which one wins?  Why do we have to construct this redundant information anyway?  Something like `_format=json` would be preferable.  But if you're really going to claim to be the framework that embraces web standards, you should support `accept` headers.

Let people handle caching in their own way.  It would still be possible to use other methods to fix the caching problem even if you did support `accept` headers.

Also, even just having a way to make an `action` always return JSON regardless of whether there is a `default` export in that file would be nice too.  Maybe call it `apiAction` or something.  Then you wouldn't be tripped up with it suddenly changing behavior when you add a `default` export later.

**Reply by sergiodxa** [reply]
The `apiAction` is any `action` in a file without a default export. That's how you build API routes.

With Single Fetch `_data` is being removed (it was always an internal thing not intended for outside usage), instead you will now add `.data` at the end of the file, so if you have a route with an action at `/users` that creates a new user you can do `POST /users.data` and _I think_ it will return just the JSON (actually it's not JSON but turbo-stream response).

**Reply by nickretallack** [reply]
That's alright.  But it's still annoying.  The more I think about it, the more I want Remix to just figure out whether you want the html or json automatically.  A simple rule could be:
* if `*/html` is in the accept types, send HTML.
* If it isn't, send the data.  Leave it as an exercise to the developer if they want further content negotiation from here.

The current behavior, and even the `.data` approach, pander far too much to the HTML use case, even though that use case should be extremely rare.  You would never need an HTML response to a POST request unless you turned off JavaScript in your browser, which is not something most people are going to do, and many web app designs just aren't going to be possible to work very well without some JavaScript.  99% of the time your POST is going to be requesting JSON from the same URL that the user is currently viewing, and Remix will be internally working around its limitation by adding `_data` or `.data`, but when you try to make the same request manually, you're likely to forget that you have to do that.  Why can't I just use let it use the `form.action` value unchanged?

The argument that this is about caching seems irrelevant to me.  Why would you cache POST responses?  Also, if your application needs full page caching, you'll need to architect around that anyway.  Many apps send different content depending on the user's session, or show otherwise unique and fast-updating content that can't be cached.  Does Remix have a recipe for that?  Can you cache part of the page and then fill in the user-specific part later?

**Reply by sergiodxa** [reply]
The reason they use `.data` instead of Accept header is because CDN cache, some (really popular) CDN ignore HTTP headers for cache, so if your CDN cache `/something` and that URL can return JSON or HTML it will cache one and return it always, by adding `.data` or `?_data` it will avoid avoid this issue since now the URL is different.

**Reply by nickretallack** [reply]
I was frustrated to find that when I enable v3 types the `.data` suffix now gives me `turbo-stream` when the server returns JSON, but remix doesn't seem to expose a way to directly access or parse this from a POST response.  `fetcher.submit` doesn't return data and `fetcher.load` doesn't accept a `method`.  Is there any way to access the parser?  Do I have to watch `useActionData`?

It's wild how many hoops I have to go through just to get the data that I'm returning from the server.  I'm not even trying to do an API here, by the way.  I just want to submit a form, get the result, and send that result to an external API (it's `Telegram.WebApp.switchInlineQuery` in case you're curious).  It seems my options are:

* Put the action at a different route from the form that submits to it.
* Get the response with `useActionData` and watch that with a `useEffect` to send it to the second step, triggering a needless re-render in the process.
* Pass the secret `_data` parameter in v2, or find a way to parse `turbo-stream` in v3.

Or, remix could support automatic content negotiation and it could just work with no changes to the developer interface.

**Reply by nickretallack** [reply]
Btw, which CDNs are you talking about that don't support the Accept header?  Is it just Microsoft Azure?  Can we make a special mode just for them?  :P

Regardless, nobody should be caching POST responses, since you must hit the server for those.  I guess POST is affected by this policy for consistency with GET?

**Reply by sergiodxa** [reply]
Cloudflare, Cloudflare doesn't support the `Vary` header needed to tell it to cache a response based on the Accept header. And Cloudflare is probably the most popular one.

Netlify CDN uses a custom `Netlify-Vary`, and it even ignores search params by default, you need to configure it to make `/foo?bar=1` and `/foo?bar=2` different keys in their CDN.

**Reply by nickretallack** [reply]
Hm.  I see.  It takes custom workers to make it consider the `Accept` header.

And I see that the `.data` approach also avoids using querystring arguments that might cause problems with other CDNs.

As for the `.data` approach, would you consider exposing some way to do a `fetch("thing.data", {method: "POST"})` that would parse the `turbo-stream` POST response data the same way that `useActionData` does?  I really just want to fetch data from the server without having to do an extra re-render and watch that data with `useEffect`, and it seems unfortunate to sometimes make things into "resource routes" when Remix is happily consuming the data from non-resource routes all the time, but just making it inconvenient to access that data.  Having `fetcher.submit()` return a promise for the response data instead of `void` would be one way to provide that.

**Comment by brophdawg11** [comment]
This discussion is being closed as part of the Remix v3 repository cleanup.

Remix v1/v2 was [upstreamed into React Router v7](https://remix.run/blog/wake-up-remix), so if this proposal is still relevant, please reopen or continue the discussion in the [React Router](https://github.com/remix-run/react-router) repository.
